spring-boot/src/main/java/org/springframework/boot/web/support/ErrorPageFilter.java

```diff
@@ -223,12 +223,12 @@ private String getErrorPath(Class<?> type) {
 		if (this.subtypes.containsKey(type)) {
 			return this.exceptions.get(this.subtypes.get(type));
 		}
-		Class<?> subtype = type;
-		while (subtype != Object.class) {
-			subtype = subtype.getSuperclass();
-			if (this.exceptions.containsKey(subtype)) {
-				this.subtypes.put(subtype, type);
-				return this.exceptions.get(subtype);
+		Class<?> supertype = type;
+		while (supertype != Object.class) {
+			supertype = supertype.getSuperclass();
+			if (this.exceptions.containsKey(supertype)) {
+				this.subtypes.put(type, supertype);
```

Comment:
Just to be clear, I was saying that the current code and tests are bad, not your PR.

Given that it would appear that the optimisation is faulty in its current state, and no one has raised it as a performance problem, I'm very tempted to remove it and simplify the code. If the optimisation is worth having then it's worth testing. To do that, the filter will need to be refactored. I would create a separate package-private class that's sole responsibility is mapping from a status or exception to an error path and implement it in such a way that it can be spied on or whatever to verify that the optimisation is working as intended.